ext/spl/spl_observer.c

```diff
@@ -361,6 +436,58 @@ PHP_METHOD(SplObjectStorage, attach)
 	spl_object_storage_attach(intern, obj, inf);
 } /* }}} */
 
+static int spl_object_storage_has_dimension(zend_object *object, zval *offset, int check_empty)
+{
+	spl_SplObjectStorage *intern = spl_object_storage_from_obj(object);
+	if (UNEXPECTED(offset == NULL || Z_TYPE_P(offset) != IS_OBJECT || (intern->flags & SOS_OVERRIDDEN_READ_DIMENSION))) {
+		/* Can't optimize empty()/isset() check if getHash, offsetExists, or offsetGet is overridden */
+		return zend_std_has_dimension(object, offset, check_empty);
+	}
+	spl_SplObjectStorageElement *element = zend_hash_index_find_ptr(&intern->storage, Z_OBJ_HANDLE_P(offset));
+	if (!element) {
+		return 0;
+	}
+
+	if (check_empty) {
+		return i_zend_is_true(&element->inf);
+	}
+	/* NOTE: SplObjectStorage->offsetExists() is an alias of SplObjectStorage->contains(), so this returns true even if the value is null. */
+	return 1;
+}
+
+static zval *spl_object_storage_read_dimension(zend_object *object, zval *offset, int type, zval *rv)
+{
+	spl_SplObjectStorage *intern = spl_object_storage_from_obj(object);
+	if (UNEXPECTED(offset == NULL || Z_TYPE_P(offset) != IS_OBJECT || (intern->flags & SOS_OVERRIDDEN_READ_DIMENSION))) {
+		/* Can't optimize it if getHash, offsetExists, or offsetGet is overridden */
+		return zend_std_read_dimension(object, offset, type, rv);
+	}
+	spl_SplObjectStorageElement *element = zend_hash_index_find_ptr(&intern->storage, Z_OBJ_HANDLE_P(offset));
+	if (type == BP_VAR_IS) {
+		if (element == NULL) {
+			return &EG(uninitialized_zval);
+		}
+	}
+
+	if (!element) {
+		zend_throw_exception_ex(spl_ce_UnexpectedValueException, 0, "Object not found");
+		return NULL;
+	} else {
+		ZVAL_COPY_DEREF(rv, &element->inf);
```

Comment:
Is it actually possible to place a reference in the object storage? I'd expect not via offsetGet, but the write dim handler might need a deref on assignment?

Reply:
I guess these handlers would cause a change in behavior for SplObjectStorage.

I probably should exclude handler `types` corrresponding to creating a reference?

```
php > $spl = new WeakMap();
php > $o = new stdClass();
php > $spl = new SplObjectStorage();
php > $o = new stdClass();
php > $spl[$o] = null;
php > $a = &$spl[$o];

Notice: Indirect modification of overloaded element of SplObjectStorage has no effect in php shell code on line 1
```


```
php > $wm = new WeakMap();
php > $o = new stdClass();
php > $wm[$o] = null;
php > $a = &$wm[$o];
php > $a = 123;
php > var_dump($wm);
object(WeakMap)#1 (1) {
  [0]=>
  array(2) {
    ["key"]=>
    object(stdClass)#2 (0) {
    }
    ["value"]=>
    &int(123)
  }
}

```

Reply:
Seems like it's impossible.